### PR TITLE
Support Java, Go, Ruby, and Rust in project-depends-on patterns

### PR DESCRIPTION
**Pull request by mmcqd**
This PR adds support for 4 new languages to `r2c-internal-project-depends-on` patterns.
* Ruby
    * Dependencies scanned from `Gemfile.lock`
* Rust
    * Dependencies scanned from `cargo.lock`  
* Go
    * Dependencies scanned from `go.sum`. Pseudo versions and `+incompatible` versions are ignored, meaning that `v3.0.1-0.20171022003610-9aa49832a739+incompatible` is treated as just `3.0.1`. If a dependency is listed both as and as not a `go module`, then both dependencies are included separately
* Java
    * Dependencies scanned from `pom.xml`. Maven does not actually create lockfiles, so instead we scan the dependency manifest, and include dependencies which are listed with an exact version. References to the `<properties>` field via `${property}` are supported.

This PR also adds `defusedxml` as a new semgrep dependency, to allow the safe parsing of XML files. This is necessary for supporting Java.

Test plan: make test (added e2e tests for each new language)

See PR to update docs: https://github.com/returntocorp/semgrep-docs/pull/465

PR checklist:

- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -45,6 +45,7 @@ repos:
           - click~=8.0
           - packaging~=21.0
           - requests~=2.22
+          - defusedxml~=0.7.1
```
Comment (aryx) [inline .pre-commit-config.yaml:48]: I remember there was some discussions on security vulnerabilities in this library. Or was it for another dependency?
Very meta :)

File: semgrep/dependencyparser/find_lockfiles.py [hunk 2]
```diff
@@ -6,7 +6,15 @@
 from typing import Optional
 from typing import Set
 
-TARGET_LOCKFILE_FILENAMES = ["pipfile.lock", "yarn.lock", "package-lock.json"]
+TARGET_LOCKFILE_FILENAMES = [
+    "pipfile.lock",
```
Comment (aryx) [inline semgrep/dependencyparser/find_lockfiles.py]: suggestion: Maybe we could add in comment the language this covers? Just to get a sense of how many languages this feature supports.

File: semgrep/dependencyparser/models.py [hunk 3]
```diff
@@ -7,16 +7,22 @@
 
 from attrs import frozen
 
-KNOWN_HASH_ALGORITHMS: Dict[str, Callable] = {
+KNOWN_HASH_ALGORITHMS: Dict[str, Optional[Callable]] = {
```
Comment (aryx) [inline semgrep/dependencyparser/models.py]: suggestion: Could you add a comment saying what this list is for? Why do we need to run those hashing functions?

File: semgrep/dependencyparser/find_lockfiles.py (hunk 2 above)
Comment (aryx) [reply semgrep/dependencyparser/find_lockfiles.py]: Or maybe we should just associate the function to parse them in a dictionary so we don't have to do those if elif later.

File: semgrep/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4shell.java/results.json [hunk 4]
```diff
@@ -0,0 +1,102 @@
+{
+  "errors": [],
+  "paths": {
+    "_comment": "<add --verbose for a list of skipped paths>",
```
Comment (aryx) [inline semgrep/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4shell.java/results.json:4]: Those elements are annoying in test files. Why do we have those _comment?

**Review by aryx — APPROVED** [review]
Fantastic!

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mmcqd) [reply .pre-commit-config.yaml:48]: The discussion was about the builtin python XML library, which has several security vulnerabilities. `defusedxml` is the library that the python docs recommend for safe XML parsing.

File: semgrep/dependencyparser/models.py (hunk 3 above)
Comment (mmcqd) [reply semgrep/dependencyparser/models.py]: We don't need to run them as far as I can tell. I was just keeping the code how @ievans set it up. I think we could probably just replace the functions with strings.

File: semgrep/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4shell.java/results.json (hunk 4 above)
Comment (mmcqd) [reply semgrep/tests/e2e/snapshots/test_dependency_aware_rules/test_dependency_aware_rules/rulesdependency_awarelog4shell.yaml-dependency_awarelog4shell.java/results.json:4]: I don't know. When I made the PR I got that github-actions request to accept the snapshot changes, and I think it added those.

File: semgrep/dependencyparser/find_lockfiles.py (hunk 2 above)
Comment (mmcqd) [reply semgrep/dependencyparser/find_lockfiles.py]: Yeah this would be nicer

**Pull request merged**